contracts/gov/proposals/TokenSaleProposal.sol

```diff
@@ -0,0 +1,349 @@
+        }
+    }
+
+    function getSaleTokenAmount(
+        uint256 tierId,
+        address tokenToBuyWith,
+        uint256 amount
+    ) public view ifTierExists(tierId) returns (uint256) {
+        require(amount > 0, "TSP: zero amount");
+
+        require(balanceOf(msg.sender, tierId) == 1, "TSP: not whitelisted");
+
+        Tier storage tier = _tiers[tierId];
+
+        TierView memory tierView = tier.tierView;
+        TierInfo storage tierInfo = tier.tierInfo;
+
+        require(
+            tierView.saleStartTime <= block.timestamp && block.timestamp <= tierView.saleEndTime,
+            "TSP: cannot buy now"
+        );
+        require(tierInfo.customers[msg.sender].purchaseTime == 0, "TSP: cannot buy twice");
+
+        uint256 exchangeRate = tierInfo.rates[tokenToBuyWith];
+        require(exchangeRate != 0, "TSP: incorrect token");
+
+        uint256 saleTokenAmount = amount.ratio(exchangeRate, PRECISION);
+
+        require(
+            tierView.minAllocationPerUser <= saleTokenAmount &&
+                saleTokenAmount <= tierView.maxAllocationPerUser,
+            "TSP: wrong allocation"
+        );
+
+        require(
+            tierInfo.totalSold + saleTokenAmount <= tierView.totalTokenProvided,
+            "TSP: insufficient sale token amount"
+        );
+
+        return saleTokenAmount;
+    }
+
+    function getVestingWithdrawAmounts(
+        uint256[] calldata tierIds
+    ) public view returns (uint256[] memory vestingWithdrawAmounts) {
+        vestingWithdrawAmounts = new uint256[](tierIds.length);
+
+        for (uint256 i = 0; i < tierIds.length; i++) {
+            vestingWithdrawAmounts[i] = _getVestingWithdrawAmount(tierIds[i]);
+        }
+    }
+
+    function getRecoveringAmounts(
+        uint256[] calldata tierIds
+    ) public view returns (uint256[] memory recoveringAmounts) {
+        recoveringAmounts = new uint256[](tierIds.length);
+
+        for (uint256 i = 0; i < recoveringAmounts.length; i++) {
+            recoveringAmounts[i] = _getRecoveringAmount(tierIds[i]);
+        }
+    }
+
+    function getTiers(
+        uint256 offset,
+        uint256 limit
+    ) external view returns (TierView[] memory tierViews) {
+        uint256 to = (offset + limit).min(latestTierId).max(offset);
+
+        tierViews = new TierView[](to - offset);
+
+        for (uint256 i = offset; i < to; i++) {
+            tierViews[i - offset] = _tiers[i + 1].tierView;
+        }
+    }
+
+    function _createTier(TierView calldata tierView) internal {
+        require(tierView.saleTokenAddress != address(0), "TSP: sale token cannot be zero");
+        require(tierView.saleTokenAddress != ETHEREUM_ADDRESS, "TSP: cannot sale native currency");
+
+        require(
+            tierView.saleStartTime <= tierView.saleEndTime,
+            "TSP: saleEndTime is less than saleStartTime"
+        );
+
+        require(
+            tierView.minAllocationPerUser <= tierView.maxAllocationPerUser &&
```

Comment:
We should also support the case when these values are set to 0, which will mean unlimited amounts

Reply:
done